### ISSUE TITLE
Tree shaking & code splitting on Web

### ISSUE THREAD

**Issue by itsezc** [DioxusLabs/dioxus#1191]
## Specific Demand

This is a high level issue (for tracking purposes) - for the web platform, when developing the binary size for WASM files get really big upwards of 100 MB for around 10 components (of course, this varies on what exactly you're building) and while there are ways to optimise this, the wasm binary is still shipping all of the app in one go, which makes for poor performance.

## Implement Suggestion

Instead, Dioxus should split the code based on the route the user requests and perform tree shaking, perhaps via #1020 to reduce the bundle size as much as possible before performing any WASM level optimisations. This should also be the case for #1183 when built to SSG, only the required WASM should be shipped.

**Comment by DogeDark** [comment]
Decided to check the change in size between Dioxus 0.2.4 and Dioxus Master using my company website (the full release built dist folder). These are the results:
```
Dioxus 0.2.4 - 866KB
Dioxus Master - 994KB

Increase: 128KB
```
The only dependency for this website (besides Dioxus) is chrono, so it doesn't reflect the true size of Dioxus itself, only the increase is reflected correctly. Both were built with ``trunk build --release``

P.S. Brotli compression is able to compress a roughly 750KB Dioxus WASM file to 265KB. A whopping 485KB decrease.

**Comment by attilio-oliva** [comment]
For a simple single route unfinished project, with few dependency calls, I got similar results using the current master branch. Using `dioxus build --release` with LTO and strip enabled, plus  `codegen-units = 1` the wasm output size is about 979 KiB.Using wasm-opt I managed to get it to around 731 KiB (~230 KiB if gzipped). But it is still quite huge for a dummy single page website with just 4 button and a modal. It does not even bind properly all the required dependencies for the project to work, since I was just trying out dioxus. 

Inspecting the wasm result using [twiggy](https://github.com/rustwasm/twiggy) on the original one, I get about 30% of the bundle size from `__wasm_bindgen_unstable`, and in the wasm-opt version it rises up to 40%.
So, the biggest problem I guess is around the wasm format itself not having a way to access the actual DOM and have to rely on bindgen. Then, code splitting would probably mitigate this for app with lots of routes, because users would download more little file instead of a single big one file that blocks the first draw, especially on mobile users. But, the structural issue won't go away unless a proposal around WASM DOM access is accepted and integrated.

Anyway, there is something odd going on because even with other similiar wasm project, dioxus bundle size output [is the biggest around](https://krausest.github.io/js-framework-benchmark/2023/table_chrome_114.0.5735.90.html). 
![image](https://github.com/DioxusLabs/dioxus/assets/75045331/44c3e3a9-dfe6-42de-90a2-0a83463482d5)
 Therefore, I must have missed something. Maybe dioxus do more operations on the actual DOM than the others, or is someway related to its architecture around the VDOM.

**Comment by Demonthos** [comment]
Dioxus uses an event system that is built for rust. Because of this, it ends up shipping more code than you would normally with something like Leptos that uses the browser event system. This allows us to make events work across platform and makes the event API nicer (for example the FileEngine API).

https://github.com/DioxusLabs/dioxus/issues/1163 should let us snip parts of the event system that are not used in your application which will improve bundle sizes.

We also have a custom WASM-bindgen like library that Dioxus uses called `slegehammer-bindgen`. From what I have tested, the bundle sizes from that library are significantly smaller than wasm-bindgen. I could add more features to that library that would allow us to switch more of the bindings over from wasm-bindgen

**Comment by itsezc** [comment]
Closed infavour of #1578
